File: fraiseql-clojure/test/fraiseql/security_test.clj
(ns fraiseql.security-test
  (:require [clojure.test :refer :all]
            [fraiseql.security :as security]))

;; ============================================================================
;; AUTHORIZATION TESTS (11 tests)
;; ============================================================================

(deftest test-simple-authorization-rule
  (let [config (-> (security/authorize-builder)
                   (assoc :rule "isOwner($context.userId, $field.ownerId)")
                   (assoc :description "Ownership check")
                   security/authorize)]
    (is (= "isOwner($context.userId, $field.ownerId)" (:rule config)))
    (is (= "Ownership check" (:description config)))))

(deftest test-authorization-with-policy
  (let [config (-> (security/authorize-builder)
                   (assoc :policy "ownerOnly")
                   (assoc :description "References named policy")
                   security/authorize)]
    (is (= "ownerOnly" (:policy config)))))

(deftest test-fluent-chaining
  (let [config (-> (security/authorize-builder)
                   (assoc :rule "hasPermission($context)")
                   (assoc :description "Complex rule")
                   (assoc :error-message "Access denied")
                   (assoc :recursive true)
                   (assoc :operations "read")
                   security/authorize)]
    (is (= "hasPermission($context)" (:rule config)))
    (is (= true (:recursive config)))
    (is (= "read" (:operations config)))))

(deftest test-caching-configuration
  (let [config (-> (security/authorize-builder)
                   (assoc :rule "checkAccess($context)")
                   (assoc :cacheable true)
                   (assoc :cache-duration-seconds 600)
                   security/authorize)]
    (is (= true (:cacheable config)))
    (is (= 600 (:cache-duration-seconds config)))))

(deftest test-error-message
  (let [config (-> (security/authorize-builder)
                   (assoc :rule "adminOnly($context)")
                   (assoc :error-message "Only administrators can access this")
                   security/authorize)]
    (is (= "Only administrators can access this" (:error-message config)))))

(deftest test-recursive
  (let [config (-> (security/authorize-builder)
                   (assoc :rule "checkNested($context)")
                   (assoc :recursive true)
                   (assoc :description "Applied to nested types")
                   security/authorize)]
    (is (= true (:recursive config)))))

(deftest test-operation-specific
  (let [config (-> (security/authorize-builder)
                   (assoc :rule "canDelete($context)")
                   (assoc :operations "delete")
                   (assoc :description "Only applies to delete operations")
                   security/authorize)]
    (is (= "delete" (:operations config)))))

(deftest test-to-map
  (let [config (-> (security/authorize-builder)
                   (assoc :rule "testRule")
                   (assoc :description "Test")
                   security/authorize)
        map-result (security/authorize-config->map config)]
    (is (= "testRule" (:rule map-result)))
    (is (= "Test" (:description map-result)))))

(deftest test-multiple-configurations
  (let [config1 (security/authorize-config :rule "rule1")
        config2 (security/authorize-config :rule "rule2")]
    (is (not= (:rule config1) (:rule config2)))))

(deftest test-default-cache-settings
  (let [config (security/authorize-config :rule "test")]
    (is (= true (:cacheable config)))
    (is (= 300 (:cache-duration-seconds config)))))

(deftest test-all-options
  (let [config (-> (security/authorize-builder)
                   (assoc :rule "complex")
                   (assoc :policy "policy")
                   (assoc :description "Complex authorization")
                   (assoc :error-message "Error")
                   (assoc :recursive true)
                   (assoc :operations "create,read,update,delete")
                   (assoc :cacheable false)
                   (assoc :cache-duration-seconds 1000)
                   security/authorize)]
    (is (= "complex" (:rule config)))
    (is (= false (:cacheable config)))
    (is (= 1000 (:cache-duration-seconds config)))))

;; ============================================================================
;; ROLE BASED ACCESS CONTROL TESTS (18 tests)
;; ============================================================================

(deftest test-single-role-requirement
  (let [config (security/role-required-config :roles ["admin"])]
    (is (= 1 (count (:roles config))))
    (is (= "admin" (first (:roles config))))))

(deftest test-multiple-role-requirements
  (let [config (security/role-required-config :roles ["manager" "director"])]
    (is (= 2 (count (:roles config))))
    (is (some #{"manager"} (:roles config)))
    (is (some #{"director"} (:roles config)))))

(deftest test-any-role-strategy
  (let [config (security/role-required-config
                 :roles ["viewer" "editor"]
                 :strategy security/ROLE_MATCH_ANY)]
    (is (= security/ROLE_MATCH_ANY (:strategy config)))))

(deftest test-all-role-strategy
  (let [config (security/role-required-config
                 :roles ["admin" "auditor"]
                 :strategy security/ROLE_MATCH_ALL)]
    (is (= security/ROLE_MATCH_ALL (:strategy config)))))

(deftest test-exactly-role-strategy
  (let [config (security/role-required-config
                 :roles ["exact_role"]
                 :strategy security/ROLE_MATCH_EXACTLY)]
    (is (= security/ROLE_MATCH_EXACTLY (:strategy config)))))

(deftest test-role-hierarchy
  (let [config (security/role-required-config :roles ["admin"] :hierarchy true)]
    (is (= true (:hierarchy config)))))

(deftest test-role-inheritance
  (let [config (security/role-required-config :roles ["editor"] :inherit true)]
    (is (= true (:inherit config)))))

(deftest test-operation-specific-roles
  (let [config (security/role-required-config :roles ["editor"] :operations "create,update")]
    (is (= "create,update" (:operations config)))))

(deftest test-role-error-message
  (let [config (security/role-required-config
                 :roles ["admin"]
                 :error-message "Administrator access required")]
    (is (= "Administrator access required" (:error-message config)))))

(deftest test-role-caching
  (let [config (security/role-required-config
                 :roles ["viewer"]
                 :cacheable true
                 :cache-duration 1800)]
    (is (= true (:cacheable config)))
    (is (= 1800 (:cache-duration-seconds config)))))

(deftest test-admin-pattern
  (let [config (security/role-required-config
                 :roles ["admin"]
                 :strategy security/ROLE_MATCH_ANY
                 :description "Admin access")]
    (is (= 1 (count (:roles config))))
    (is (= "admin" (first (:roles config))))))

(deftest test-manager-director-pattern
  (let [config (security/role-required-config
                 :roles ["manager" "director"]
                 :strategy security/ROLE_MATCH_ANY)]
    (is (= 2 (count (:roles config))))
    (is (= security/ROLE_MATCH_ANY (:strategy config)))))

(deftest test-data-scientist-pattern
  (let [config (security/role-required-config
                 :roles ["data_scientist" "analyst"]
                 :strategy security/ROLE_MATCH_ANY)]
    (is (= 2 (count (:roles config))))))

(deftest test-role-to-map
  (let [config (security/role-required-config
                 :roles ["admin" "editor"]
                 :strategy security/ROLE_MATCH_ANY)
        map-result (security/role-required-config->map config)]
    (is (= "any" (:strategy map-result)))))

(deftest test-role-description
  (let [config (security/role-required-config
                 :roles ["viewer"]
                 :description "Read-only access")]
    (is (= "Read-only access" (:description config)))))

(deftest test-role-default-values
  (let [config (security/role-required-config :roles ["user"])]
    (is (= false (:hierarchy config)))
    (is (= false (:inherit config)))
    (is (= true (:cacheable config)))
    (is (= 300 (:cache-duration-seconds config)))))

;; ============================================================================
;; ATTRIBUTE BASED ACCESS CONTROL TESTS (16 tests)
;; ============================================================================

(deftest test-abac-policy-creation
  (let [config (security/authz-policy-config "accessControl"
                 :type security/AUTHZ_POLICY_ABAC
                 :attributes ["clearance_level >= 2"]
                 :description "Basic clearance")]
    (is (= "accessControl" (:name config)))
    (is (= security/AUTHZ_POLICY_ABAC (:type config)))))

(deftest test-multiple-attributes
  (let [config (security/authz-policy-config "secretAccess"
                 :type security/AUTHZ_POLICY_ABAC
                 :attributes ["clearance_level >= 3" "background_check == true"])]
    (is (= 2 (count (:attributes config))))))

(deftest test-clearance-level-policy
  (let [config (security/authz-policy-config "topSecret"
                 :type security/AUTHZ_POLICY_ABAC
                 :attributes ["clearance_level >= 3"])]
    (is (= 1 (count (:attributes config))))))

(deftest test-department-policy
  (let [config (security/authz-policy-config "financeDept"
                 :type security/AUTHZ_POLICY_ABAC
                 :attributes ["department == \"finance\""])]
    (is (= "financeDept" (:name config)))))

(deftest test-time-based-policy
  (let [config (security/authz-policy-config "businessHours"
                 :type security/AUTHZ_POLICY_ABAC
                 :attributes ["now >= 9:00 AM" "now <= 5:00 PM"])]
    (is (= 2 (count (:attributes config))))))

(deftest test-geographic-policy
  (let [config (security/authz-policy-config "usOnly"
                 :type security/AUTHZ_POLICY_ABAC
                 :attributes ["country == \"US\""])]
    (is (= 1 (count (:attributes config))))))

(deftest test-gdpr-policy
  (let [config (security/authz-policy-config "gdprCompliance"
                 :type security/AUTHZ_POLICY_ABAC
                 :attributes ["gdpr_compliant == true" "data_residency == \"EU\""])]
    (is (= 2 (count (:attributes config))))))

(deftest test-data-classification-policy
  (let [config (security/authz-policy-config "classifiedData"
                 :type security/AUTHZ_POLICY_ABAC
                 :attributes ["classification >= 2"])]
    (is (= 1 (count (:attributes config))))))

(deftest test-abac-caching
  (let [config (security/authz-policy-config "cachedAccess"
                 :type security/AUTHZ_POLICY_ABAC
                 :attributes ["role == \"viewer\""]
                 :cacheable true
                 :cache-duration 600)]
    (is (= true (:cacheable config)))
    (is (= 600 (:cache-duration-seconds config)))))

(deftest test-abac-audit-logging
  (let [config (security/authz-policy-config "auditedAccess"
                 :type security/AUTHZ_POLICY_ABAC
                 :attributes ["audit_enabled == true"]
                 :audit-logging true)]
    (is (= true (:audit-logging config)))))

(deftest test-abac-recursive
  (let [config (security/authz-policy-config "recursiveAccess"
                 :type security/AUTHZ_POLICY_ABAC
                 :attributes ["permission >= 1"]
                 :recursive true)]
    (is (= true (:recursive config)))))

(deftest test-abac-operation-specific
  (let [config (security/authz-policy-config "readOnly"
                 :type security/AUTHZ_POLICY_ABAC
                 :attributes ["can_read == true"]
                 :operations "read")]
    (is (= "read" (:operations config)))))

(deftest test-complex-abac-policy
  (let [config (security/authz-policy-config "complex"
                 :type security/AUTHZ_POLICY_ABAC
                 :attributes ["level >= 2" "verified == true" "active == true"]
                 :audit-logging true
                 :cacheable true)]
    (is (= 3 (count (:attributes config))))
    (is (= true (:audit-logging config)))))

(deftest test-abac-error-message
  (let [config (security/authz-policy-config "restricted"
                 :type security/AUTHZ_POLICY_ABAC
                 :attributes ["clearance >= 2"]
                 :error-message "Insufficient clearance level")]
    (is (= "Insufficient clearance level" (:error-message config)))))

(deftest test-abac-to-map
  (let [config (security/authz-policy-config "test"
                 :type security/AUTHZ_POLICY_ABAC
                 :attributes ["test >= 1"])
        map-result (security/authz-policy-config->map config)]
    (is (= "abac" (:type map-result)))))

(deftest test-abac-default-values
  (let [config (security/authz-policy-config "default"
                 :type security/AUTHZ_POLICY_ABAC)]
    (is (= true (:cacheable config)))
    (is (= 300 (:cache-duration-seconds config)))
    (is (= false (:recursive config)))))

;; ============================================================================
;; AUTHORIZATION POLICY TESTS (19 tests)
;; ============================================================================

(deftest test-rbac-policy
  (let [config (security/authz-policy-config "adminOnly"
                 :type security/AUTHZ_POLICY_RBAC
                 :rule "hasRole($context, 'admin')"
                 :description "Access restricted to administrators"
                 :audit-logging true)]
    (is (= "adminOnly" (:name config)))
    (is (= security/AUTHZ_POLICY_RBAC (:type config)))
    (is (= "hasRole($context, 'admin')" (:rule config)))
    (is (= true (:audit-logging config)))))

(deftest test-abac-policy-full
  (let [config (security/authz-policy-config "secretClearance"
                 :type security/AUTHZ_POLICY_ABAC
                 :description "Requires top secret clearance"
                 :attributes ["clearance_level >= 3" "background_check == true"])]
    (is (= "secretClearance" (:name config)))
    (is (= security/AUTHZ_POLICY_ABAC (:type config)))
    (is (= 2 (count (:attributes config))))))

(deftest test-custom-policy
  (let [config (security/authz-policy-config "customRule"
                 :type security/AUTHZ_POLICY_CUSTOM
                 :rule "isOwner($context.userId, $resource.ownerId)")]
    (is (= security/AUTHZ_POLICY_CUSTOM (:type config)))))

(deftest test-hybrid-policy
  (let [config (security/authz-policy-config "auditAccess"
                 :type security/AUTHZ_POLICY_HYBRID
                 :rule "hasRole($context, 'auditor')"
                 :attributes ["audit_enabled == true"])]
    (is (= security/AUTHZ_POLICY_HYBRID (:type config)))
    (is (= "hasRole($context, 'auditor')" (:rule config)))))

(deftest test-multiple-policies
  (let [p1 (security/authz-policy-config "policy1" :type security/AUTHZ_POLICY_RBAC)
        p2 (security/authz-policy-config "policy2" :type security/AUTHZ_POLICY_ABAC)
        p3 (security/authz-policy-config "policy3" :type security/AUTHZ_POLICY_CUSTOM)]
    (is (= "policy1" (:name p1)))
    (is (= "policy2" (:name p2)))
    (is (= "policy3" (:name p3)))))

(deftest test-pii-access-policy
  (let [config (security/authz-policy-config "piiAccess"
                 :type security/AUTHZ_POLICY_RBAC
                 :rule "hasRole($context, 'data_manager')")]
    (is (= "piiAccess" (:name config)))))

(deftest test-admin-only-policy
  (let [config (security/authz-policy-config "adminOnly"
                 :type security/AUTHZ_POLICY_RBAC
                 :audit-logging true)]
    (is (= true (:audit-logging config)))))

(deftest test-recursive-policy
  (let [config (security/authz-policy-config "recursiveProtection"
                 :type security/AUTHZ_POLICY_CUSTOM
                 :recursive true)]
    (is (= true (:recursive config)))))

(deftest test-operation-specific-policy
  (let [config (security/authz-policy-config "readOnly"
                 :type security/AUTHZ_POLICY_CUSTOM
                 :operations "read")]
    (is (= "read" (:operations config)))))

(deftest test-cached-policy
  (let [config (security/authz-policy-config "cachedAccess"
                 :type security/AUTHZ_POLICY_CUSTOM
                 :cacheable true
                 :cache-duration 3600)]
    (is (= true (:cacheable config)))
    (is (= 3600 (:cache-duration-seconds config)))))

(deftest test-audited-policy
  (let [config (security/authz-policy-config "auditedAccess"
                 :type security/AUTHZ_POLICY_RBAC
                 :audit-logging true)]
    (is (= true (:audit-logging config)))))

(deftest test-policy-with-error-message
  (let [config (security/authz-policy-config "restrictedAccess"
                 :type security/AUTHZ_POLICY_RBAC
                 :error-message "Only executive level users can access this resource")]
    (is (= "Only executive level users can access this resource" (:error-message config)))))

(deftest test-policy-fluent-chaining
  (let [builder (-> (security/authz-policy-builder "complexPolicy")
                    (assoc :type security/AUTHZ_POLICY_HYBRID)
                    (assoc :rule "hasRole($context, 'admin')")
                    (assoc :attributes ["security_clearance >= 3"])
                    (assoc :cacheable true)
                    (assoc :cache-duration-seconds 1800)
                    (assoc :recursive false)
                    (assoc :operations "create,update,delete")
                    (assoc :audit-logging true)
                    (assoc :error-message "Insufficient privileges"))
        config (security/build-policy builder)]
    (is (= "complexPolicy" (:name config)))
    (is (= security/AUTHZ_POLICY_HYBRID (:type config)))
    (is (= true (:cacheable config)))
    (is (= true (:audit-logging config)))))

(deftest test-policy-composition
  (let [p1 (security/authz-policy-config "publicAccess" :type security/AUTHZ_POLICY_RBAC :rule "true")
        p2 (security/authz-policy-config "piiAccess" :type security/AUTHZ_POLICY_RBAC)
        p3 (security/authz-policy-config "adminAccess" :type security/AUTHZ_POLICY_RBAC)]
    (is (= "publicAccess" (:name p1)))
    (is (= "piiAccess" (:name p2)))
    (is (= "adminAccess" (:name p3)))))

(deftest test-financial-data-policy
  (let [config (security/authz-policy-config "financialData"
                 :type security/AUTHZ_POLICY_ABAC
                 :attributes ["clearance_level >= 2" "department == \"finance\""])]
    (is (= "financialData" (:name config)))
    (is (= 2 (count (:attributes config))))))

(deftest test-security-clearance-policy
  (let [config (security/authz-policy-config "secretClearance"
                 :type security/AUTHZ_POLICY_ABAC
                 :attributes ["clearance_level >= 3" "background_check == true"])]
    (is (= 2 (count (:attributes config))))))

(deftest test-default-policy
  (let [config (security/authz-policy-config "default")]
    (is (= "default" (:name config)))
    (is (= security/AUTHZ_POLICY_CUSTOM (:type config)))
    (is (= true (:cacheable config)))
    (is (= 300 (:cache-duration-seconds config)))))

(deftest test-policy-to-map
  (let [config (security/authz-policy-config "test"
                 :type security/AUTHZ_POLICY_RBAC
                 :rule "test_rule")
        map-result (security/authz-policy-config->map config)]
    (is (= "test" (:name map-result)))
    (is (= "rbac" (:type map-result)))))

File: fraiseql-elixir/test/fraiseql/security_test.exs
defmodule FraiseQL.SecurityTest do
  use ExUnit.Case
  doctest FraiseQL.Security

  alias FraiseQL.Security

  # ===== Authorization Tests (11 tests) =====

  describe "authorize_config" do
    test "creates simple rule" do
      config = Security.authorize_config(rule: "test_rule")
      assert config.rule == "test_rule"
      assert config.policy == ""
      assert config.cacheable == true
    end

    test "uses policy reference" do
      config = Security.authorize_config(policy: "admin_policy")
      assert config.policy == "admin_policy"
      assert config.rule == ""
    end

    test "builder pattern with authorize/1" do
      builder = Security.authorize_builder()
      |> Map.put(:rule, "test_rule")
      |> Map.put(:description, "Test description")

      config = Security.authorize(builder)
      assert config.rule == "test_rule"
      assert config.description == "Test description"
    end

    test "caching configuration" do
      config = Security.authorize_config(
        rule: "test",
        cacheable: true,
        cache_duration_seconds: 600
      )
      assert config.cacheable == true
      assert config.cache_duration_seconds == 600
    end

    test "error message configuration" do
      config = Security.authorize_config(
        rule: "test",
        error_message: "Access denied"
      )
      assert config.error_message == "Access denied"
    end

    test "recursive application" do
      config = Security.authorize_config(
        rule: "test",
        recursive: true
      )
      assert config.recursive == true
    end

    test "operation-specific rules" do
      config = Security.authorize_config(
        rule: "test",
        operations: "read,write"
      )
      assert config.operations == "read,write"
    end

    test "defaults to cacheable=true and 300 seconds" do
      config = Security.authorize_config()
      assert config.cacheable == true
      assert config.cache_duration_seconds == 300
    end

    test "all configuration options" do
      config = Security.authorize_config(
        rule: "isOwner(...)",
        policy: "owner_policy",
        description: "Owner check",
        error_message: "Not owner",
        recursive: true,
        operations: "read",
        cacheable: false,
        cache_duration_seconds: 0
      )
      assert config.rule == "isOwner(...)"
      assert config.policy == "owner_policy"
      assert config.description == "Owner check"
      assert config.error_message == "Not owner"
      assert config.recursive == true
      assert config.operations == "read"
      assert config.cacheable == false
      assert config.cache_duration_seconds == 0
    end

    test "equality comparison" do
      config1 = Security.authorize_config(rule: "test")
      config2 = Security.authorize_config(rule: "test")
      assert config1 == config2
    end
  end

  # ===== Role-Based Access Control Tests (18 tests) =====

  describe "role_required_config" do
    test "single role requirement" do
      config = Security.role_required_config(roles: ["admin"])
      assert config.roles == ["admin"]
      assert config.strategy == :any
    end

    test "multiple roles" do
      config = Security.role_required_config(roles: ["manager", "director"])
      assert config.roles == ["manager", "director"]
    end

    test "role matching strategy: any" do
      config = Security.role_required_config(
        roles: ["admin", "manager"],
        strategy: :any
      )
      assert config.strategy == :any
      assert Security.strategy_to_string(:any) == "any"
    end

    test "role matching strategy: all" do
      config = Security.role_required_config(
        roles: ["admin", "auditor"],
        strategy: :all
      )
      assert config.strategy == :all
      assert Security.strategy_to_string(:all) == "all"
    end

    test "role matching strategy: exactly" do
      config = Security.role_required_config(
        roles: ["admin"],
        strategy: :exactly
      )
      assert config.strategy == :exactly
      assert Security.strategy_to_string(:exactly) == "exactly"
    end

    test "role hierarchy" do
      config = Security.role_required_config(
        roles: ["admin"],
        hierarchy: true
      )
      assert config.hierarchy == true
    end

    test "role inheritance" do
      config = Security.role_required_config(
        roles: ["user"],
        inherit: true
      )
      assert config.inherit == true
    end

    test "operation-specific requirements" do
      config = Security.role_required_config(
        roles: ["admin"],
        operations: "delete"
      )
      assert config.operations == "delete"
    end

    test "custom error message" do
      config = Security.role_required_config(
        roles: ["admin"],
        error_message: "Admin access required"
      )
      assert config.error_message == "Admin access required"
    end

    test "admin pattern" do
      config = Security.role_required_config(
        roles: ["admin"],
        description: "Admin only access"
      )
      assert config.roles == ["admin"]
      assert config.description == "Admin only access"
    end

    test "manager pattern" do
      config = Security.role_required_config(
        roles: ["manager", "director"],
        strategy: :any,
        description: "Manager or director"
      )
      assert config.roles == ["manager", "director"]
      assert config.strategy == :any
    end

    test "data scientist pattern" do
      config = Security.role_required_config(
        roles: ["data_scientist", "analyst"],
        strategy: :any
      )
      assert config.roles == ["data_scientist", "analyst"]
    end

    test "caching configuration" do
      config = Security.role_required_config(
        roles: ["admin"],
        cacheable: true,
        cache_duration_seconds: 1800
      )
      assert config.cacheable == true
      assert config.cache_duration_seconds == 1800
    end

    test "builder pattern with build_roles/1" do
      builder = Security.role_required_builder()
      |> Map.put(:roles, ["admin", "manager"])
      |> Map.put(:strategy, :any)

      config = Security.build_roles(builder)
      assert config.roles == ["admin", "manager"]
      assert config.strategy == :any
    end

    test "all configuration options" do
      config = Security.role_required_config(
        roles: ["admin", "manager"],
        strategy: :any,
        hierarchy: true,
        description: "Leadership",
        error_message: "Leader required",
        operations: "write",
        inherit: true,
        cacheable: false,
        cache_duration_seconds: 0
      )
      assert config.roles == ["admin", "manager"]
      assert config.strategy == :any
      assert config.hierarchy == true
      assert config.description == "Leadership"
      assert config.error_message == "Leader required"
      assert config.operations == "write"
      assert config.inherit == true
      assert config.cacheable == false
      assert config.cache_duration_seconds == 0
    end

    test "equality comparison" do
      config1 = Security.role_required_config(roles: ["admin"])
      config2 = Security.role_required_config(roles: ["admin"])
      assert config1 == config2
    end
  end

  # ===== Attribute-Based Access Control Tests (16 tests) =====

  describe "authz_policy_config with ABAC" do
    test "ABAC policy definition" do
      config = Security.authz_policy_config("secret_access",
        type: :abac,
        attributes: ["clearance_level >= 3"]
      )
      assert config.type == :abac
      assert config.attributes == ["clearance_level >= 3"]
    end

    test "multiple attributes" do
      config = Security.authz_policy_config("complex",
        type: :abac,
        attributes: ["department == 'engineering'", "tenure >= 2"]
      )
      assert Enum.count(config.attributes) == 2
    end

    test "clearance level checking" do
      config = Security.authz_policy_config("top_secret",
        type: :abac,
        attributes: ["clearance_level >= 5", "background_check == true"]
      )
      assert config.attributes == ["clearance_level >= 5", "background_check == true"]
    end

    test "department-based access" do
      config = Security.authz_policy_config("hr_only",
        type: :abac,
        attributes: ["department == 'human_resources'"]
      )
      assert config.attributes == ["department == 'human_resources'"]
    end

    test "time-based access control" do
      config = Security.authz_policy_config("business_hours",
        type: :abac,
        attributes: ["current_hour >= 9", "current_hour < 17"]
      )
      assert Enum.count(config.attributes) == 2
    end

    test "geographic restrictions" do
      config = Security.authz_policy_config("geo_limited",
        type: :abac,
        attributes: ["country == 'US'"]
      )
      assert config.attributes == ["country == 'US'"]
    end

    test "GDPR compliance" do
      config = Security.authz_policy_config("gdpr_compliant",
        type: :abac,
        attributes: ["region == 'EU'", "data_processing == true"]
      )
      assert config.attributes == ["region == 'EU'", "data_processing == true"]
    end

    test "data classification" do
      config = Security.authz_policy_config("pii_access",
        type: :abac,
        attributes: ["classification == 'public'"]
      )
      assert config.attributes == ["classification == 'public'"]
    end

    test "caching with TTL" do
      config = Security.authz_policy_config("cached_abac",
        type: :abac,
        attributes: ["test"],
        cacheable: true,
        cache_duration_seconds: 3600
      )
      assert config.cacheable == true
      assert config.cache_duration_seconds == 3600
    end

    test "audit logging" do
      config = Security.authz_policy_config("audited",
        type: :abac,
        attributes: ["test"],
        audit_logging: true
      )
      assert config.audit_logging == true
    end

    test "recursive attribute application" do
      config = Security.authz_policy_config("recursive_abac",
        type: :abac,
        attributes: ["test"],
        recursive: true
      )
      assert config.recursive == true
    end

    test "operation-specific attributes" do
      config = Security.authz_policy_config("write_restricted",
        type: :abac,
        attributes: ["test"],
        operations: "write"
      )
      assert config.operations == "write"
    end

    test "complex attribute combinations" do
      config = Security.authz_policy_config("complex_abac",
        type: :abac,
        attributes: ["a == 1", "b == 2", "c == 3"]
      )
      assert Enum.count(config.attributes) == 3
    end

    test "custom error message" do
      config = Security.authz_policy_config("error_abac",
        type: :abac,
        attributes: ["test"],
        error_message: "Insufficient attributes"
      )
      assert config.error_message == "Insufficient attributes"
    end

    test "equality comparison" do
      config1 = Security.authz_policy_config("test",
        type: :abac,
        attributes: ["a"]
      )
      config2 = Security.authz_policy_config("test",
        type: :abac,
        attributes: ["a"]
      )
      assert config1 == config2
    end
  end

  # ===== Authorization Policy Tests (19 tests) =====

  describe "authz_policy_config for all types" do
    test "RBAC policy type" do
      config = Security.authz_policy_config("admin_policy",
        type: :rbac,
        rule: "hasRole($context, 'admin')"
      )
      assert config.type == :rbac
      assert Security.policy_type_to_string(:rbac) == "rbac"
    end

    test "ABAC policy type" do
      config = Security.authz_policy_config("attribute_policy",
        type: :abac
      )
      assert config.type == :abac
      assert Security.policy_type_to_string(:abac) == "abac"
    end

    test "CUSTOM policy type" do
      config = Security.authz_policy_config("custom_policy",
        type: :custom,
        rule: "custom_expression"
      )
      assert config.type == :custom
      assert Security.policy_type_to_string(:custom) == "custom"
    end

    test "HYBRID policy type" do
      config = Security.authz_policy_config("hybrid_policy",
        type: :hybrid,
        rule: "hybrid_rule",
        attributes: ["attr1"]
      )
      assert config.type == :hybrid
      assert Security.policy_type_to_string(:hybrid) == "hybrid"
    end

    test "multiple policies" do
      config1 = Security.authz_policy_config("policy1")
      config2 = Security.authz_policy_config("policy2")
      assert config1.name == "policy1"
      assert config2.name == "policy2"
    end

    test "PII access policy" do
      config = Security.authz_policy_config("piiAccess",
        type: :rbac,
        rule: "hasRole($context, 'data_manager')",
        description: "PII data access"
      )
      assert config.name == "piiAccess"
      assert config.description == "PII data access"
    end

    test "admin-only policy" do
      config = Security.authz_policy_config("adminOnly",
        type: :rbac,
        rule: "hasRole($context, 'admin')"
      )
      assert config.name == "adminOnly"
      assert config.rule == "hasRole($context, 'admin')"
    end

    test "recursive policy application" do
      config = Security.authz_policy_config("recursive_policy",
        recursive: true
      )
      assert config.recursive == true
    end

    test "operation-specific policies" do
      config = Security.authz_policy_config("delete_policy",
        operations: "delete"
      )
      assert config.operations == "delete"
    end

    test "cached policies" do
      config = Security.authz_policy_config("cached_policy",
        cacheable: true,
        cache_duration_seconds: 3600
      )
      assert config.cacheable == true
      assert config.cache_duration_seconds == 3600
    end

    test "audited policies" do
      config = Security.authz_policy_config("audited_policy",
        audit_logging: true
      )
      assert config.audit_logging == true
    end

    test "custom error messages" do
      config = Security.authz_policy_config("error_policy",
        error_message: "Policy denied"
      )
      assert config.error_message == "Policy denied"
    end

    test "policy composition" do
      policy1 = Security.authz_policy_config("p1", type: :rbac)
      policy2 = Security.authz_policy_config("p2", type: :abac)
      assert policy1.type == :rbac
      assert policy2.type == :abac
    end

    test "builder pattern with build_policy/1" do
      builder = Security.authz_policy_builder("test_policy")
      |> Map.put(:type, :rbac)
      |> Map.put(:rule, "test_rule")

      config = Security.build_policy(builder)
      assert config.name == "test_policy"
      assert config.type == :rbac
      assert config.rule == "test_rule"
    end

    test "fluent builder chaining" do
      builder = Security.authz_policy_builder("chain_policy")
      |> Map.put(:type, :hybrid)
      |> Map.put(:rule, "rule1")
      |> Map.put(:attributes, ["attr1"])
      |> Map.put(:description, "Chained")

      config = Security.build_policy(builder)
      assert config.type == :hybrid
      assert config.rule == "rule1"
      assert config.attributes == ["attr1"]
      assert config.description == "Chained"
    end

    test "financial data policy" do
      config = Security.authz_policy_config("financialData",
        type: :rbac,
        rule: "hasRole($context, 'finance_manager')",
        audit_logging: true
      )
      assert config.name == "financialData"
      assert config.audit_logging == true
    end

    test "security clearance policy" do
      config = Security.authz_policy_config("securityClearance",
        type: :abac,
        attributes: ["clearance_level >= 3"]
      )
      assert config.name == "securityClearance"
      assert config.attributes == ["clearance_level >= 3"]
    end

    test "default configuration" do
      config = Security.authz_policy_config("default_policy")
      assert config.name == "default_policy"
      assert config.type == :custom
      assert config.description == ""
      assert config.rule == ""
      assert config.attributes == []
      assert config.cacheable == true
      assert config.cache_duration_seconds == 300
      assert config.recursive == false
      assert config.operations == ""
      assert config.audit_logging == false
      assert config.error_message == ""
    end

    test "equality comparison" do
      config1 = Security.authz_policy_config("test")
      config2 = Security.authz_policy_config("test")
      assert config1 == config2
    end
  end
end
